tests/FSharp.Compiler.ComponentTests/Conformance/Constraints/Unmanaged.fs

```diff
@@ -485,6 +501,71 @@ printf "%s" (CsharpStruct<int>.Hi<MultiCaseUnion>())
     IL_0000:  ret
   } """]
 
+    [<Fact>]
+    let ``FSharp does not generate modreq for VBNET to consume in v8`` () = 
+        Fsx "let testMyFunction (x: 'TUnmanaged when 'TUnmanaged : unmanaged) = ()"
+        |> withLangVersion80
+        |> compile
+        |> shouldSucceed
+        |> verifyIL ["""
+      .method public static void  testMyFunction<TUnmanaged>(!!TUnmanaged x) cil managed
+  {
+    
+    .maxstack  8
+    IL_0000:  ret
+  } """]
+
+    [<Fact>]
+    let ``Unmanaged constraint in lambda reproduces issue 17509`` () = 
+        // This test reproduces the issue https://github.com/dotnet/fsharp/issues/17509
+        // When UnmanagedConstraintCsharpInterop is enabled, it generates invalid IL
+        // causing a TypeLoadException at runtime
+        Fsx """
+open System
+
+let printTypeConstraintsNative<'T when 'T : unmanaged> () = printf $"Hello: {typeof<'T>.FullName} is unmanaged"
+
+let Main() =
+    let func (x:int) : 'T when 'T : unmanaged = Unchecked.defaultof<'T>
+    let initFinite = Seq.init<nativeint> 3 func
+    printfn "%A" initFinite
+ 
+printTypeConstraintsNative<nativeint>()
+Main()
+        """
+        |> withLangVersionPreview
+        |> asExe
+        |> compile
+        |> shouldSucceed
+
+    [<FactForNETCOREAPP>]
+    let ``Unmanaged constraint in lambda generates invalid IL for Specialize method with preview version`` () = 
+        Fsx """
+let Main() =
+    let func (x:int) : 'T when 'T : unmanaged = Unchecked.defaultof<'T>
+    let initFinite = Seq.init<nativeint> 3 func
+    printfn "%A" initFinite
+Main()
+        """
+        |> withLangVersionPreview
+        |> asExe
+        |> compile
+        |> shouldSucceed
+        |> verifyIL ["""
+    .method public strict virtual instance object Specialize<valuetype (class [runtime]System.ValueType modreq([runtime]System.Runtime.InteropServices.UnmanagedType)) T>() cil managed
```

Comment:
Expand this IL block to show the full contents of the generated `func@3` class as well.
You first need to change this string to something like `xyz`, run it, get actual IL, and paste here a bigger block showing the full context.